File: dart/dartfind/lib/src/file_util.dart
import 'dart:io' show Platform;

import 'package:path/path.dart' as path;
import 'package:dartfind/src/find_exception.dart';

class FileUtil {
  static final dotDirs = {'.', '..'};

  static String extension(String filePath) {
    var ext = path.extension(filePath);
    if (ext.isNotEmpty && ext.startsWith('.')) {
      ext = ext.substring(1);
    }
    return ext;
  }

  static String homePath() {
    if (Platform.environment.containsKey('HOME')) {
      return path.absolute(Platform.environment['HOME']!);
    } else if (Platform.environment.containsKey('USERPROFILE')) {
      return path.absolute(Platform.environment['USERPROFILE']!);
    }
    throw FindException('Home path not found');
  }

  static String expandPath(String filePath) {
    if (filePath.startsWith('~')) {
      var userPath = FileUtil.homePath();
      if (filePath == '~' || filePath == '~${Platform.pathSeparator}') {
        return userPath;
      } else if (filePath.startsWith('~${Platform.pathSeparator}')) {
        return path.join(userPath, filePath.substring(2));
      }
      // Another user's home directory
      var homePath = path.dirname(userPath);
      return path.join(homePath, filePath.substring(1));
    }
    return filePath;
  }

  static String contractPath(String filePath) {
    var homePath = FileUtil.homePath();
    if (filePath.startsWith(homePath)) {
      return filePath.replaceFirst(homePath, '~');
    }
    return filePath;
  }

  static List<String> splitPath(String filePath) {
    return path.split(filePath).where((e) => e.isNotEmpty).toList();
  }

  static Set<String> pathElems(String filePath) {
    return path.split(filePath).where((e) => e.isNotEmpty).toSet();
  }

  static bool isDotDir(String filePath) => dotDirs.contains(filePath);

  static bool isHiddenName(String name) {
    return name.length > 1 && name.startsWith('.') && !isDotDir(name);
  }

  static bool isHiddenPath(String filePath) {
    var elems = pathElems(filePath);
    return elems.any((elem) => isHiddenName(elem));
  }

  static String normalizePath(String filePath) {
    if (filePath.endsWith(Platform.pathSeparator)) {
      return filePath.substring(0, filePath.length - 2);
    }
    return filePath;
  }

  static int sepCount(String filePath) {
    var normPath = normalizePath(filePath);
    return Platform.pathSeparator.allMatches(normPath).length;
  }
}

File: go/gofind/pkg/gofind/fileutil.go
package gofind

import (
	"fmt"
	"os"
	"os/user"
	"path/filepath"
	"strings"
)

const (
	Dot    = "."
	DotDot = ".."
	Tilde  = "~"
)

func ExpandPath(filePath string) string {
	if strings.HasPrefix(filePath, Tilde) {
		userPath := getHome()
		tildeSlash := fmt.Sprintf("~%c", os.PathSeparator)
		if filePath == Tilde || filePath == tildeSlash {
			return userPath
		}
		if strings.HasPrefix(filePath, tildeSlash) {
			return filepath.Join(userPath, filePath[2:])
		}
		// Another user's home directory
		homePath := filepath.Dir(userPath)
		return filepath.Join(homePath, filePath[1:])
	}
	return filePath
}

func GetExtension(file string) string {
	ext := filepath.Ext(filepath.Base(file))
	return strings.ToLower(strings.TrimLeft(ext, Dot))
}

func getHome() string {
	usr, err := user.Current()
	if err != nil || usr == nil {
		if err != nil {
			LogError(err.Error())
		}
		return ""
	}
	return usr.HomeDir
}

func getPathSeparatorString() string {
	return fmt.Sprintf("%c", os.PathSeparator)
}

func isDotDir(file string) bool {
	dotDirs := []string{Dot, DotDot}
	return containsV(dotDirs, file)
}

func IsHiddenName(name string) bool {
	return len(name) > 1 && strings.HasPrefix(name, Dot) && !isDotDir(name)
}

func IsHiddenPath(path string) bool {
	for _, elem := range strings.Split(path, getPathSeparatorString()) {
		if IsHiddenName(elem) {
			return true
		}
	}
	return false
}

func normalizePath(path string) string {
	return strings.TrimRight(path, "/\\")
}

func relativePath(path string, startPath string) string {
	homePath := getHome()
	Log(fmt.Sprintf("homePath:%s", homePath))
	relativePath := path
	if startPath == Dot && strings.HasPrefix(path, homePath) {
		Log("path starts with homePath")
		relativePath = Dot + strings.TrimPrefix(path, homePath)
	}
	return relativePath
}
